File: LeetCode/Binary Tree Preorder Traversal/java/src/BinaryTreePreOrderTraversal.java
import java.util.LinkedList;
import java.util.List;

public class BinaryTreePreOrderTraversal {

    public static List<Integer> InOrderTraversalIterative (TreeNode root) {

        LinkedList<Integer> result = new LinkedList();
        LinkedList<TreeNode> stack = new LinkedList();

        stack.add(root);

        if (root == null) {
            return result;
        }

        while(!stack.isEmpty()) {

            TreeNode node = stack.pollLast();

            result.add(node.val);

            if(node.left != null) {
                stack.add(node.left);
            }

            if(node.right != null) {
                stack.add(node.right);
            }

        }

        return result;
    }

    // Instantiate LinkedList every recursive call
    public static List<Integer> InOrderTraversalRecursive (TreeNode root) {

        List<Integer> result = new LinkedList<Integer>();
        if (root != null){
            result.add(root.val);
            result.addAll(InOrderTraversalRecursive(root.left));
            result.addAll(InOrderTraversalRecursive(root.right));
        }
        return result;
    }

    // Instantiate LinkedList Once
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> pre = new LinkedList<Integer>();
        preHelper(root,pre);
        return pre;
    }

    public void preHelper(TreeNode root, List<Integer> pre) {
        if(root==null) return;
        pre.add(root.val);
        preHelper(root.left,pre);
        preHelper(root.right,pre);
    }

    public static void main(String[] args) {

        TreeNode node1 = new TreeNode(1);
        TreeNode node2 = new TreeNode(2);
        TreeNode node3 = new TreeNode(3);

        node1.right = node2;
        node2.left = node3;

        List<Integer> result = InOrderTraversalIterative(node1);

        for (Integer nodeVal : result) {
            System.out.println(nodeVal);
        }

        node1 = new TreeNode(1);
        node2 = new TreeNode(2);
        node3 = new TreeNode(3);

        node1.right = node2;
        node2.left = node3;

        List<Integer> result2 = InOrderTraversalRecursive(node1);

        for (Integer nodeVal : result2) {
            System.out.println(nodeVal);
        }

    }

}

File: LeetCode/Binary Tree Preorder Traversal/cpp/binary_tree_preorder_traversal.cpp
#include <bits/stdc++.h>

using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    vector<int> preOrder;
    vector<int> res;
    
public:
    vector<int> preorderTraversalIterative(TreeNode *root) {
        stack<TreeNode*> nodes;
        
        if(root == NULL) //in case the input is empty
            return preOrder;
        
        nodes.push(root);
        
        while(!nodes.empty()){
            TreeNode *tmpNode = nodes.top();
            
            //psuh the root's value into a vector
            preOrder.push_back(tmpNode -> val);
            
            nodes.pop();//pop the tmpNode from stack
            
            //because statk is LIFO, the right node is latter than left node
            if(tmpNode -> right != NULL)    nodes.push(tmpNode -> right);
            if(tmpNode -> left  != NULL)    nodes.push(tmpNode -> left);
        }
        
        return preOrder;
    }

    void preorderTraversalRecursiveHelper(TreeNode* root, vector<int> &res) {
        if(root) {
            res.push_back(root->val);           

            if(root->left)
                preorderTraversalRecursiveHelper(root->left, res);

            if(root->right)
                preorderTraversalRecursiveHelper(root->right, res);
        }

        return;
    }

    vector<int> preorderTraversalRecursive(TreeNode* root) {   
        preorderTraversalRecursiveHelper(root, res) ;
        return res;
    }

};

int main() {

    // Make the nodes
    TreeNode node1(1);
    TreeNode node2(2);
    TreeNode node3(3);
    
    // Connect nodes
    node1.right = &node2;
    node2.left = &node3;

    Solution solution;
    vector<int> result1 = solution.preorderTraversalIterative(&node1);

    for(auto res: result1) {
        cout << res << endl;
    }

    vector<int> result2 = solution.preorderTraversalRecursive(&node1);

    for(auto res: result2) {
        cout << res << endl;
    }

    return 0;
}